### PR TITLE
[Refactor] Create visitor for dotExp method of type AST nodes

### PR DESCRIPTION
**Pull request by RazvanN7**
cc @WalterBright

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
This doesn't need to be exposed in the C++ headers?

**Comment by RazvanN7** [comment]
>This doesn't need to be exposed in the C++ headers?

When I created visitors for the semantic methods I didn't have to so I assumed that's fine not to in this case.

**Review by wilzbach — APPROVED** [review]
Can we please go with `dotexp.d` if we already move code around?
(ideally all of these would be in a `visitors` package, but well seems like [this](https://github.com/dlang/dmd/pull/7639) ain't happening)

**Comment by JinShil** [comment]
@RazvanN7 Please rebase if you want this to move forwarrd.

**Comment by RazvanN7** [comment]
@JinShil done, although I don't know why it fails on Darwin

**Comment by jacob-carlborg** [comment]
If you only moved some code you might have missed the changes that the latest objc PR did to in the `dotExp` method. This PR was recently merged.

**Comment by jacob-carlborg** [comment]
@RazvanN7 This code needs to be included: https://github.com/dlang/dmd/blob/bf824c6170cab5a152668fe488a87954fea07415/src/dmd/mtype.d#L8411-L8419

**Comment by RazvanN7** [comment]
@jacob-carlborg thank you very much! Done.

File: src/dmd/typesem.d [hunk 1]
```diff
@@ -2543,3 +2548,1371 @@ private extern(C++) final class ResolveVisitor : Visitor
...
+ *  resulting expression with e.ident resolved
+ */
+extern(C++) Expression dotExp(Type mt, Scope* sc, Expression e, Identifier ident, int flag)
+{
+    scope v = new DotExpVisitor(sc, e, ident, flag);
+    mt.accept(v);
+    return v.result;
+}
+
+private extern(C++) final class DotExpVisitor : Visitor
```
Comment (andralex) [inline src/dmd/typesem.d]: Make extern(D) since it's not private anyway

File: src/dmd/typesem.d [hunk 2]
```diff
@@ -2543,3 +2548,1371 @@ private extern(C++) final class ResolveVisitor : Visitor
...
+ *  mt = type for which the dot expression is used
+ *  sc = instantiating scope
+ *  e = expression to convert
+ *  ident = identifier being used
+ *  flag = DotExpFlag bit flags
+ *
+ * Returns:
+ *  resulting expression with e.ident resolved
+ */
+extern(C++) Expression dotExp(Type mt, Scope* sc, Expression e, Identifier ident, int flag)
```
Comment (andralex) [inline src/dmd/typesem.d]: cc @ibuclaw does the C++ interface need this at all? If not let's make it `extern(D)` for better modularity.

**Review by andralex — APPROVED** [review]
I'd need to understand better what the C++ interface needs and what it doesn't so we minimize the surface. cc @ibuclaw @JohanEngelen @klickverbot @redstar

**Pull request merged**